### PR TITLE
Replacing invalid xml character by unicode escape character

### PR DESCRIPTION
**Pull request by Faizan2304**
Issue: 

Fix: 	Replace invalid xml character by unicode escape character

Testing:
1) UT to cover changes.
2) Manually validated by creating trx file with invalid xml

### CODE REVIEW

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs [hunk 1]
```diff
@@ -700,13 +700,23 @@ private static string RemoveInvalidXmlChar(string str)
                 // From xml spec (http://www.w3.org/TR/xml/#charsets) valid chars: 
                 // #x9 | #xA | #xD | [#x20-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF]  
 
-                string invalidChar = @"[^\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u10000-\u10FFFF]";
-                return Regex.Replace(str, invalidChar, TrxLoggerResources.PlaceHolderForInvalidChar);
+                // we are handling only #x9 | #xA | #xD | [#x20-#xD7FF] | [#xE000-#xFFFD]
```
Comment (sbaid) [inline src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs]: add unit test string to include Unicode character valid & invalid in all of these ranges

File: test/Microsoft.TestPlatform.Extensions.TrxLogger.UnitTests/XmlPersistenceTests.cs [hunk 2]
```diff
@@ -13,14 +13,15 @@ public class XmlPersistenceTests
         [TestMethod]
         public void SaveObjectShouldRemoveInvalidCharacter()
         {
+            System.Diagnostics.Debugger.Launch();
```
Comment (sbaid) [inline test/Microsoft.TestPlatform.Extensions.TrxLogger.UnitTests/XmlPersistenceTests.cs]: need to remove this

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs [hunk 3]
```diff
@@ -700,13 +700,23 @@ private static string RemoveInvalidXmlChar(string str)
...
+                string invalidChar = @"[^\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD]";
+                return Regex.Replace(str, invalidChar, evaluator);
             }
 
             return str;
         }
 
+        private static string ReplaceInvalidCharacterWithUniCodeEscapeSequence(Match match)
+        {
+            char x = match.Value.ToArray()[0];
```
Comment (sbaid) [inline src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs]: match.Value[0] can be used

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs (hunk 1 above)
Comment (sbaid) [inline src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs]: do a perf run on a 10k, 1L tests on utperf06 to assess the impact.

Comment (Faizan2304) [inline src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs]: addressed in next commit

File: test/Microsoft.TestPlatform.Extensions.TrxLogger.UnitTests/XmlPersistenceTests.cs (hunk 2 above)
Comment (Faizan2304) [inline test/Microsoft.TestPlatform.Extensions.TrxLogger.UnitTests/XmlPersistenceTests.cs]: Removed!!

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs (hunk 3 above)
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs]: Yes we can use this.

**Comment by Faizan2304** [comment]
Testing:
1) Covered code change with unit test.
2) Manually validated that invalid xml char are getting replace by escape sequence and trx is getting loaded using XmlDocument.

**Comment by Faizan2304** [comment]
By default invalid xml character will not get replace. To replace it with escape sequence use `` --logger:trx;EscapeInvalidChar=true``

**Comment by AbhitejJohn** [comment]
This fix looks more of a workaround that is offloaded to the customer. There would also have discoverability issues trying to figure out this setting. Should we not do this by default if its known to be an invalid character?
@Faizan2304 : Please share the perf impact.

**Comment by Faizan2304** [comment]
image


From the data it seems that replacing invalid xml is not affecting much on perf.

File: src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs (hunk 1 above)
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.Extensions.TrxLogger/XML/XmlPersistence.cs]: Done. Please find the result below comment.
